### PR TITLE
conan.io support

### PR DESCRIPTION
**Pull request by Manu343726**

### CODE REVIEW

**Comment by foonathan** [comment]
Your test package stuff does not seem to work.

Also I don't like that you have to copy an example for it.

**Comment by Manu343726** [comment]
> Your test package stuff does not seem to work.

Yeah, I think is a bug in the cmake generator of conan. I've sent a PR to fix it https://github.com/conan-io/conan/pull/801

> Also I don't like that you have to copy an example for it.

I agree it is not the best way to handle this, but I can also create the example cmake target using the examples/optional.cpp file directly.

**Comment by foonathan** [comment]
> I agree it is not the best way to handle this, but I can also create the example cmake target using the examples/optional.cpp file directly.

Yes, that would be better.

**Comment by foonathan** [comment]
Ping @Manu343726.

**Comment by Manu343726** [comment]
It seems there's still something weird with their cmake macro to check the compiler version:https://github.com/conan-io/conan/issues/1011

We have a C++ meetup in Madrid in a couple of weeks, so reaching @drodri to (finally) fix this would be easier... (I need to review this with him since I don't see something specially weird in the code they are generating nor in the code generator)

*NOTE: I'm explicitly spamming him since it seems this is the only way to make him not to ignore my messages :P*

**Comment by foonathan** [comment]
Ping?

**Comment by foonathan** [comment]
So, how's it going?

File: .travis.yml [hunk 1]
```diff
@@ -65,13 +65,16 @@ install:
...
+  - pip install --user conan && export PATH=$PATH:$HOME/.local/bin
+
   - cd type_safe/
 
 script:
   - mkdir build/ && cd build/
   - $CMAKE -DCMAKE_BUILD_TYPE=$BUILD_TYPE -DCMAKE_CXX_FLAGS="-Wall -Wextra -pedantic -Wno-parentheses" -DTYPE_SAFE_BUILD_TEST_EXAMPLE=ON -DTYPE_SAFE_BUILD_DOC="$BUILD_DOCS" -DTYPE_SAFE_STANDARDESE_DOWNLOAD_DIRECTORY=../standardese/ ../
   - $CMAKE --build .
   - ./test/type_safe_test
+  - cd .. && conan test_package --build=outdated && cd build
```
Comment (memsharded) [inline .travis.yml]: This is very outdated now, please use ``conan create``

File: README.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 [![Build Status](https://travis-ci.org/foonathan/type_safe.svg?branch=master)](https://travis-ci.org/foonathan/type_safe)
 [![Build status](https://ci.appveyor.com/api/projects/status/aw1j2h2s52g4laen/branch/master?svg=true)](https://ci.appveyor.com/project/foonathan/type-safe/branch/master)
+[![badge](https://img.shields.io/badge/conan.io-type_safe%2F0.1-green.svg?logo=data:image/png;base64%2CiVBORw0KGgoAAAANSUhEUgAAAA4AAAAOCAMAAAAolt3jAAAA1VBMVEUAAABhlctjlstkl8tlmMtlmMxlmcxmmcxnmsxpnMxpnM1qnc1sn85voM91oM11oc1xotB2oc56pNF6pNJ2ptJ8ptJ8ptN9ptN8p9N5qNJ9p9N9p9R8qtOBqdSAqtOAqtR%2BrNSCrNJ/rdWDrNWCsNWCsNaJs9eLs9iRvNuVvdyVv9yXwd2Zwt6axN6dxt%2Bfx%2BChyeGiyuGjyuCjyuGly%2BGlzOKmzOGozuKoz%2BKqz%2BOq0OOv1OWw1OWw1eWx1eWy1uay1%2Baz1%2Baz1%2Bez2Oe02Oe12ee22ujUGwH3AAAAAXRSTlMAQObYZgAAAAFiS0dEAIgFHUgAAAAJcEhZcwAACxMAAAsTAQCanBgAAAAHdElNRQfgBQkREyOxFIh/AAAAiklEQVQI12NgAAMbOwY4sLZ2NtQ1coVKWNvoc/Eq8XDr2wB5Ig62ekza9vaOqpK2TpoMzOxaFtwqZua2Bm4makIM7OzMAjoaCqYuxooSUqJALjs7o4yVpbowvzSUy87KqSwmxQfnsrPISyFzWeWAXCkpMaBVIC4bmCsOdgiUKwh3JojLgAQ4ZCE0AMm2D29tZwe6AAAAAElFTkSuQmCC)](http://www.conan.io/source/type_safe/0.1/Manu343726/testing)
```
Comment (memsharded) [inline README.md]: These badges are outdated too, now using Bintray

File: conanfile.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import os
+
+class TypeSafe(ConanFile):
+    name = 'type_safe'
+    url  = 'https://foonathan.github.io/blog/2016/10/11/type-safe.html'
+    version = '0.1'
+    username = 'Manu343726'
+    channel = 'testing'
+    requires = 'debug_assert/1.1@Manu343726/testing'
+    exports = '*.hpp'
```
Comment (memsharded) [inline conanfile.py]: Use ``exports_sources`` instead

File: conanfile.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+from conans import ConanFile
+from conans.tools import download, untargz
+import os
+
+class TypeSafe(ConanFile):
+    name = 'type_safe'
+    url  = 'https://foonathan.github.io/blog/2016/10/11/type-safe.html'
```
Comment (memsharded) [inline conanfile.py]: The URL should point to the repo containing the recipe. It is the entry point to collaborate and give feedback.

File: conanfile.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    version = '0.1'
+    username = 'Manu343726'
```
Comment (memsharded) [inline conanfile.py]: username and channel shouldn't be in the recipe at all

File: test_package/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from conans import ConanFile, CMake
+import os
+
+class TypeSafe(ConanFile):
+    settings = 'os', 'compiler', 'build_type', 'arch'
+    requires = (
+        'cmake-utils/0.0.0@Manu343726/testing',
+        'type_safe/0.1@Manu343726/testing'
```
Comment (memsharded) [inline test_package/conanfile.py]: This requirement can be removed now, it is autoamtically injected

File: test_package/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    )
+    generators = 'cmake'
+
+    def build(self):
+        cmake = CMake(self.settings)
+        self.run('cmake {} {}'.format(self.conanfile_directory, cmake.command_line))
```
Comment (memsharded) [inline test_package/conanfile.py]: new ``cmake.configure()`` and ``cmake.build()`` methods might be more convenient.

File: test_package/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,15 @@
+project(type_safe_test_package)
+cmake_minimum_required(VERSION 2.8)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+include(${CONAN_CMAKE-UTILS_ROOT}/conan.cmake)
```
Comment (memsharded) [inline test_package/CMakeLists.txt]: This extra include of conan.cmake might be unnecessary. You might want to update to modern CMake syntax with ``conan_basic_setup(TARGETS)``

**Comment by agauniyal** [comment]
I'd like to see this library in conan as well 😄

**Comment by Manu343726** [comment]
pong (sorry). I would like to get into conan again in a couple of weeks, this lib being the no 1 in my list. BTW I would need your help @memsharded to get into the new jecosystem

**Comment by Manu343726** [comment]
I've finally got some time to update the packages. Both type safe and debug assert packages have been updated and pushed to my artifactory repository (https://bintray.com/manu343726/conan-packages). I'm waiting for the travis CI build to see if the package test is passing.

**Comment by Manu343726** [comment]
BTW the last commit allows any to own the package with no change in the recipe, just upload the package using your account and repo:

``` shell
conan remote add your_remote_name your_bintray_repo_url # i.e. https://bintray.com/manu343726/conan-packages
conan user -p your_bintray_user_token -r your_remote_name your_user_name
conan create your_user/your_channel # this generates and tests the package in channel "your_user/your_channel"
conan upload type_safe/0.1@your_user/your_channel --all --remote your_remote_name
```

**Comment by memsharded** [comment]
If you go one step further, you can also reuse it for future versions without modifying the recipe. Remove the name & version from the recipe then use:

```
conan create . PkgName/version@user/channel
```

**Comment by Manu343726** [comment]
Oh, didn't know about that. Thanks my anonymous friend.

**Comment by Manu343726** [comment]
this last commit should make the travis CI build compatible with the conan 1.0 stable release.

File: conanfile.py (hunk 4 above)
Comment (foonathan) [reply conanfile.py]: @Manu343726 Could you fix it?

File: conanfile.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+from conans import ConanFile
+from conans.tools import download, untargz
+import os
+
+class TypeSafe(ConanFile):
+    name = 'type_safe'
+    url  = 'https://foonathan.github.io/blog/2016/10/11/type-safe.html'
+    version = '0.1'
+    requires = 'debug_assert/1.3@Manu343726/testing'
```
Comment (foonathan) [inline conanfile.py]: I'm guessing I have to update this one everytime I release a new version and/or update debug_assert?

**Comment by Manu343726** [comment]
> I don't really want to add a lot of releases for type_safe, how would that work with conan? Can people just get the latest mater?

We could maintain two different channels for the package (think of channels as "upstreams"), one with fixed package releases (the "stable" channel) and another one pointing to the latest master version of type_safe ("master"?, "latest"). Neither options are a problem from the recipe perspective (it's python after all, dwtfywt), just a matter of organization.

*I'm not sure if conan packaging allows packages without explicit versioning. @memsharded? 
Anyway, the different versions of the latest channel could just correspond to fixes in the recipes.*

> @Manu343726 Could you fix it?

Of course, sorry. I will check debug_assert recipe too just in case.

> I'm guessing I have to update this one everytime I release a new version and/or update debug_assert?

Right, but we could infer the dependency programmatically from your sources. If type_safe uses a fixed debug_assert tag as submodule, we could parse that version from type_safe sources and set the package version from that tag. Note you would need to release a debug_assert package with the required version first.
OTOH we could follow the "latest" channel approach above, and tell the debug_assert package to use the latest version tagged in the sources:

``` python
def get_package_version():
   return os.system("git describe --abbrev=0 --tags", cwd=???)

class DebugAssert(ConanFile):
    version = get_package_version()
...
```

In general, I see what you're asking for: To release new versions of type_safe & debug_assert packages automatically, taking into account users can rely on the master branch, so packages should pick the latest master sources directly. What do you thing @memsharded ? How would you handle this use case? (I'm not an expert, my comments above were just suggestions that come to my mind.

**Comment by memsharded** [comment]
In many cases you just want to have packages for tagged/published versions, in this case, v0.1.

Though you can perfectly have just ``pkg/master@...`` versions too. Packages get overwritten from the latest master.

**Comment by Manu343726** [comment]
@memsharded so we could tag the package as master, cool. About getting the package version from git or an environment variable, what do you think?

**Comment by memsharded** [comment]
You can leave the version undefined in the recipe. Then, in your CI, for example, get it from a branch convention (like release/X.Y.Z), and use:

```
$ conan create . Pkg/X.Y.Z@user/channel
```

**Comment by Manu343726** [comment]
Purrrrrfect, thanks a lot

**Comment by foonathan** [comment]
@Manu343726 Could you just update the project URL and then I'll merge it.

**Comment by Manu343726** [comment]
Sure, I was trying to find a way to automate the dependency version.

**Comment by Manu343726** [comment]
Done. Note both the package version and the dependency version would require manual changes in the recipe. Also, the debug_assert dependency is pointing to my jfrog account, it's up to you to configure your own account and submit debug_assert there.

**Pull request merged**